# Business Rules Quick Reference

## Order Creation Rules ‚úÖ
- ‚úÖ Orders must have unique, non-null ID
- ‚úÖ Quantity must be positive (> 0)
- ‚úÖ Price currency must match symbol quote currency
- ‚úÖ Orders start in PENDING status

## Order State Transitions üîÑ
```
PENDING ‚Üí PARTIAL ‚Üí FILLED ‚úÖ
PENDING ‚Üí FILLED ‚úÖ
PENDING ‚Üí CANCELLED ‚úÖ
PARTIAL ‚Üí CANCELLED ‚úÖ
PARTIAL ‚Üí FILLED ‚úÖ

‚ùå FILLED ‚Üí * (terminal)
‚ùå CANCELLED ‚Üí * (terminal)
```

## Transaction Matching Rules üîó
- ‚úÖ Buy price >= Sell price (must overlap)
- ‚úÖ Both orders must be active (not FILLED/CANCELLED)
- ‚úÖ Orders must have same symbol
- ‚úÖ Execution price: sell_price ‚â§ exec_price ‚â§ buy_price
- ‚úÖ Optimal pricing: (buy_price + sell_price) / 2

## Quantity Management üìä
- ‚úÖ Track executed quantity (cumulative)
- ‚úÖ Track remaining quantity (original - executed)
- ‚úÖ Transaction quantity ‚â§ min(buy_remaining, sell_remaining)
- ‚úÖ Multiple transactions allowed until fully executed
- ‚úÖ Auto-transition: remaining = 0 ‚Üí FILLED

## Order Modifications üìù
- ‚úÖ Price updates only if not terminal state
- ‚úÖ Cancellation only if not already terminal
- ‚úÖ Automatic timestamp updates
- ‚ùå No modifications to FILLED/CANCELLED orders

## Data Integrity üõ°Ô∏è
- ‚úÖ Null safety with clear error messages
- ‚úÖ Immutable value objects (Money, Symbol, Currency)
- ‚úÖ Defensive copying for collections
- ‚úÖ Comprehensive validation at creation

## Error Handling ‚ö†Ô∏è
- üö´ `IllegalArgumentException` for business rule violations
- üö´ `IllegalStateException` for invalid state transitions
- üö´ `NullPointerException` for null parameters
- üìù Descriptive error messages with context

## Currency Rules üí∞
- ‚úÖ Same currency required for monetary operations
- ‚úÖ Currency-specific precision handling
- ‚úÖ Appropriate rounding strategies per currency type
- ‚úÖ Display formatting respects decimal places

## Transaction History üìö
- ‚úÖ Orders maintain chronological transaction list
- ‚úÖ Sequential numbering (1, 2, 3, ...)
- ‚úÖ Immutable history (defensive copies)
- ‚úÖ Automatic timestamp management

## Key Validation Points üîç

### Before Transaction Creation
1. ‚úÖ Orders are active (not FILLED/CANCELLED)
2. ‚úÖ Orders have same symbol
3. ‚úÖ Buy price >= Sell price
4. ‚úÖ Execution price within valid range
5. ‚úÖ Transaction quantity within limits
6. ‚úÖ All parameters non-null and valid

### After Transaction Creation
1. ‚úÖ Orders updated with transaction reference
2. ‚úÖ Executed quantities incremented
3. ‚úÖ Order statuses updated based on remaining quantity
4. ‚úÖ Timestamps updated automatically

## Common Error Scenarios ‚ùå

| Scenario | Exception | Message Pattern |
|----------|-----------|-----------------|
| Null order ID | `NullPointerException` | "ID cannot be null" |
| Negative quantity | `IllegalArgumentException` | "Quantity must be positive" |
| Wrong currency | `IllegalArgumentException` | "Price currency ... does not match ..." |
| No price overlap | `IllegalArgumentException` | "Orders cannot match: buy price is less than sell price" |
| Insufficient quantity | `IllegalArgumentException` | "Transaction quantity cannot exceed ... remaining quantity" |
| Inactive order | `IllegalArgumentException` | "... order must be active (not FILLED or CANCELLED)" |
| Terminal state modification | `IllegalStateException` | "Cannot update ... of terminal order" |

## Test Coverage Summary üìã

| Rule Category | Test Class | Key Tests |
|---------------|------------|-----------|
| State Updates | `TransactionOrderStateUpdatesTest` | 6 tests |
| Quantity Tracking | `QuantityTrackingTest` | 4 tests |
| Multiple Transactions | `MultipleTransactionSupportTest` | 3 tests |
| Order Matching | `OrderMatchingRulesTest` | 4 tests |
| Transaction Ordering | `TransactionOrderingTest` | 5 tests |
| Domain Rules | `TransactionTest` | 20+ tests |
| Integration | `OrderDomainIntegrationTest` | 2 tests |

**Total: 45+ comprehensive test scenarios**

## Architecture Patterns Used üèóÔ∏è

- **Domain-Driven Design (DDD)**: Rich domain model with business logic
- **State Pattern**: Order status management with state transitions
- **Value Objects**: Immutable Money, Symbol, Currency classes
- **Factory Methods**: Static creation methods for common objects
- **Defensive Programming**: Comprehensive validation and null safety
- **Test-Driven Development**: Extensive test coverage for all rules